[WIP] Search API improvements.

poljar opened the pull request:
This patch-set simplifies the search API and adds the ability to order the results by a given field like it was suggested in #10.

The first patch removes the concept of a collector and uses an optional argument to limit the search result.

The second patch adds another optional argument to the method that adds the ability to to order the search result by an unsigned field.

While all of this is pretty straightforward the second patch has a problem. To be able to order a search result by a field the field needs to be an unsigned field that is marked as a fast field. The API already allows this when adding such a field, the problem arises if the developer fails to mark the field as a fast one. 

The search will crash the Python interpreter when it is run without a fast field. Tantivy raises a panic under such conditions and sadly the `Searcher` isn't unwind-safe so catching the panic and turning it into a `Result` and therefore turning it into a nice Python exception isn't possible.

Since crashes should be avoided under all circumstances on the Python side of things we a couple possibilities here:
- Don't let Python users create non-fast unsigned fields
- Modify Tantivy to return a Result type instead of calling panic.
- Don't expose the order by a fast field feature at all.

poljar committed:
searcher: Remove the collector concept.

This patch removes the TopDocs collector class and adds a limit
argument on the search method.

poljar committed:
searcher: Allow the search to be sorted by an unsigned field.

fulmicoton commented on src/searcher.rs:
```diff
@@ -29,20 +32,30 @@ impl Searcher {
     ///
     /// Raises a ValueError if there was an error with the search.
+    #[args(limit = 10, sort_by = "SORT_BY")]
     fn search(
```
We cannot have the matching document count with this setting though :-S

poljar replied:
By matching count you mean using the `CountCollector`? If so, yeah I'm aware, adding that would probably mean switching to the `Multicollector`. I will probably do so at the next iteration of this.

bloodbare commented:
If its useful I've needed to implement a fast search function with MultiCollector for facets with a count collector and the top collector:

https://github.com/bloodbare/tantivy-py/blob/api-change/src/searcher.rs#L37

Its a really basic aproach and these days I don't have spare time to pulish and to a proper PR so if you want to use some of it feel free.

fulmicoton committed:
Using tantivy 0.11

poljar committed:
searcher: Use a search result struct.

poljar committed:
searcher: Remove the ability to order the search result.

Ordering the search result by a field requires the field to be set up to
support this at the index creation time. If it wasn't properly set up,
such a search would crash the Python interpreter.

Until a search returns an error that we can convert to a Python
exception this feature will unlikely be supported.
